Fix schema problems

Peilonrayz opened the pull request:
In light of #68 I have updated the code to fix #57, #58, #60, #65, #67 and #68. I have added relevant tests and ensured they pass in Python 3.6 and 3.7. (I've changed the code to allow skipping of the datetime tests.)

From this I have moved all schema code into `mm` so that the marshmallow code is contained. And this is basically just a re-write of `mm`. I've updated `api` to accommodate these changes.

Otherwise only `core` has been changed. This is because marshmallow now loads schema data into the relevant model. (This can be tested against with code in #58). I'm unsure how you'd want the data to be handled, and so I chose to return early so I don't mangle with `core` too. **please verify this is the intended behavior**.

This should keep all the previous logic around options, most are in `schema` such as specifying a default, and handling `Optional`. The other special piece of logic is when creating a DC, I kept the `field_many` option. Otherwise `build_type` now builds all the types you'd previously specified, and a couple more, in a recursive form, defaulting to `fields.Field` and a warning.

---

I would like to request that we change how you handle the `dataclasses_json.mm_field` so that if the value is `None` we generate one by default. This isn't something you did previously, and so I didn't change this.

dsvensson commented:
This fixes cases like this, hooray!

```
class Bar:
 b: str

class Foo:
  a: List[Bar]
```

But if Bar contains an Enum-field, that fails with:
```
  File ".../venv/lib/python3.7/site-packages/dataclasses_json/api.py", line 76, in schema
    Schema = build_schema(cls, DataClassJsonMixin, infer_missing, partial)
  File ".../venv/lib/python3.7/site-packages/dataclasses_json/mm.py", line 113, in build_schema
    schema_ = schema(cls, mixin, infer_missing)
  File ".../venv/lib/python3.7/site-packages/dataclasses_json/mm.py", line 98, in schema
    t = build_type(type_, options, mixin, field, cls)
  File ".../venv/lib/python3.7/site-packages/dataclasses_json/mm.py", line 73, in build_type
    return inner(type_, options)
  File ".../venv/lib/python3.7/site-packages/dataclasses_json/mm.py", line 72, in inner
    return field.Field(**options)
AttributeError: 'Field' object has no attribute 'Field'
```

Peilonrayz commented:
@dsvensson Looks like an easy fix, I used `field` rather than `fields` [here](https://github.com/lidatong/dataclasses-json/pull/66/files#diff-17d6f2f251ae64dcc26ae93bf8e23974R72).

Could you provide the code that made this error?

dsvensson commented:
@Peilonrayz Oh that was easy. It was indeed just that typo. Too much redacting, but I can provide a test case later today. There's also another test case that ought to be added, for list of classes where one field contains a list of classes. That didn't work before this branch. #57 . More or less extending the last decoder test in the test...nestedsomething..py.

Peilonrayz commented:
@dsvensson Yeah I had done that with another `fields.Fields`, thought I changed all of them, after that tho... :(

I've added the test you mention already - [test][0] & [object][1]. Note, you can reproduce this without the outer most datatype being a list. It's just that the library only allows decoding lists via `.schema`, otherwise `from_json` and other code doesn't make use the schema code.

  [0]: https://github.com/lidatong/dataclasses-json/pull/66/commits/463ea01090b0d6571bb56c403f09a994e724a39c#diff-e1bc6e780c4b78fd3a4a85bf7b8ae0acR18
  [1]: https://github.com/lidatong/dataclasses-json/pull/66/commits/463ea01090b0d6571bb56c403f09a994e724a39c#diff-753922789de93c054356e2f67368b4d9R219

lidatong commented:
> In light of #68 I have updated the code to fix #57, #58, #60, #65, #67 and #68. I have added relevant tests and ensured they pass in Python 3.6 and 3.7. (I've changed the code to allow skipping of the datetime tests.)

thanks!

> From this I have moved all schema code into `mm` so that the marshmallow code is contained. And this is basically just a re-write of `mm`. I've updated `api` to accommodate these changes.
> 
> Otherwise only `core` has been changed. This is because marshmallow now loads schema data into the relevant model. (This can be tested against with code in #58). I'm unsure how you'd want the data to be handled, and so I chose to return early so I don't mangle with `core` too. **please verify this is the intended behavior**.

will address this separately

> This should keep all the previous logic around options, most are in `schema` such as specifying a default, and handling `Optional`. The other special piece of logic is when creating a DC, I kept the `field_many` option. Otherwise `build_type` now builds all the types you'd previously specified, and a couple more, in a recursive form, defaulting to `fields.Field` and a warning.
> 
> I would like to request that we change how you handle the `dataclasses_json.mm_field` so that if the value is `None` we generate one by default. This isn't something you did previously, and so I didn't change this.

yep this makes sense for a future patch

lidatong commented:
Released in [0.2.2](https://github.com/lidatong/dataclasses-json/releases/tag/v0.2.2)

Peilonrayz commented:
@lidatong It should be noted that [this line](https://github.com/lidatong/dataclasses-json/blob/master/dataclasses_json/mm.py#L72) should be changed to `return fields.Field(**options)`. As pointed out by @dsvensson otherwise Enum support won't work.

I was waiting for a test case to work against until I committed the fix.
